Issue #10817: Fix IllegalTypeCheck on Record with configured member modifiers

Rahulkhinchi03 opened the pull request:
Fixes #10817

Diff Regression config: https://gist.githubusercontent.com/Rahulkhinchi03/7b19d57cd13d1d2904ab523a61f7f300/raw/5784d380be13aec65a30e5042a18ddb25b20308a/config.xml

Rahulkhinchi03 commented:
@romani Kindly help me with the failing test cases!

strkkk commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeRecordsWithMemberModifiersSpecified.java:
```diff
@@ -0,0 +1,28 @@
+
+public record InputIllegalTypeRecordsWithMemberModifiersSpecified(
+        UUID productId,
+        @Nullable String identifier
```
please remove annotation

strkkk commented:
@Rahulkhinchi03 run build on local machine (mvn clean verify) to see what is wrong

strkkk commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeRecordsWithMemberModifiersSpecified.java:
```diff
@@ -0,0 +1,28 @@
+)
+{
+
+}
```
new line missing

Rahulkhinchi03 replied:
How can I resolve this?

strkkk replied:
it is not like this. you dont need to paste \n at the end, just have a new line

Rahulkhinchi03 replied:
warning: CRLF will be replaced by LF

Rahulkhinchi03 commented:
[ERROR]   IllegalTypeCheckTest.testValidateRecordsWithConfiguredMemberModifiersPass:68->AbstractModuleTestSupport.verifyWithInlineConfigParser:267 » Checkstyle

I am getting this error. Can you please tell me what does it mean? @strkkk

strkkk commented:
> 
> 
> [ERROR] IllegalTypeCheckTest.testValidateRecordsWithConfiguredMemberModifiersPass:68->AbstractModuleTestSupport.verifyWithInlineConfigParser:267 » Checkstyle
> 
> I am getting this error. Can you please tell me what does it mean? @strkkk

run this test locally and check output

Rahulkhinchi03 commented:
> > [ERROR] IllegalTypeCheckTest.testValidateRecordsWithConfiguredMemberModifiersPass:68->AbstractModuleTestSupport.verifyWithInlineConfigParser:267 » Checkstyle
> > I am getting this error. Can you please tell me what does it mean? @strkkk
> 
> run this test locally and check output

com.puppycrawl.tools.checkstyle.api.CheckstyleException: Config comment not specified properly in InputIllegalTypeRecordsWithMemberModifiersSpecified.java

Rahulkhinchi03 commented:
@romani Kindly review my changes!

strkkk commented:
@Rahulkhinchi03 you need to provide regression report for your fixed check. Here is manual how to do that with github actions  https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#executing-generation
also you can check how it looks like in other bugfix prs, e.g. https://github.com/checkstyle/checkstyle/pull/10922

Rahulkhinchi03 commented:
GitHub, generate report

Rahulkhinchi03 commented:
@strkkk created the Diff Regression config.

Rahulkhinchi03 commented:
Kindly review my changes! @strkkk 
I have created the Diff Regression config.

Rahulkhinchi03 commented:
@romani sir, kindly review my changes!

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java:
```diff
@@ -491,13 +491,13 @@ public void visitToken(DetailAST ast) {
         boolean result = true;
-        if (!memberModifiers.isEmpty()) {
-            final DetailAST modifiersAst = methodOrVariableDef
+        if (!memberModifiers.isEmpty() && ast.getType() != TokenTypes.RECORD_COMPONENT_DEF) {
```
It is not clear why modifier is null.
We have modifiers on record https://github.com/checkstyle/checkstyle/blob/988cbe38621df6cf079c975fc9ca87b45a3c4af9/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java#L5447

Records should have modifiers same as other types.

Rahulkhinchi03 replied:
You mean I need to change `TokenTypes.RECORD_COMPONENT_DEF -> TokenTypes.RECORD_DEF` ?

Rahulkhinchi03 replied:
@romani sir, can you please clarify what exactly I need to change?

Rahulkhinchi03 replied:
@romani  sir, can you please clarify what exactly I need to change?

romani replied:
I do not understand why we simply skip validation
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java#L5513
It can have illegal type.

Please create input with violation on record component.

Rahulkhinchi03 replied:
@romani sir,
After changing 
public record InputIllegalTypeRecordsWithMemberModifiersSpecified
        (UUID productId, String identifier) 
-->
public record InputIllegalTypeRecordsWithMemberModifiersSpecified
        (HashSet productId, String identifier) {

This is failing and It says Violation lines for   (HashSet productId, String identifier) {expected to be empty.

romani replied:
This still worries me.
I probably need to debug this code to understand it

romani replied:
please make code like:

```
            case TokenTypes.RECORD_COMPONENT_DEF:
                visitComponentDef(ast);
                break;
```

```
visitComponentDef() {
    if (isContainVerifiableType(variableDef)) {
            checkClassName(variableDef);
        }
}
```

nmancus1 replied:
@romani condition in `visitComponentDef` if statement should be extended to check `memberModifiers` for `private` and `final` modifiers. The two following type declarations should be equivalent to this check, and we should expect violations on both with default config (how it works currently) and with any combination of `FINAL` and/or `LITERAL_PRIVATE`:

```java
import java.util.TreeSet;

class TestClass {
    private final TreeSet x;

    TestClass(TreeSet x) {
        this.x = x;
    }
}

record TestRecord(TreeSet x) {}

```

Rahulkhinchi03 replied:
@romani @nmancus1
After adding Romani sir's code:

```
case TokenTypes.RECORD_COMPONENT_DEF:
                visitComponentDef(ast);
                break;
```

```
  /**
     * Checks type of ComponentDef.
     *
     * @param variableDef to check.
     */
    private void visitComponentDef(DetailAST variableDef) {
        if (isContainVerifiableType(variableDef)) {
            checkClassName(variableDef);
        }
    }
```
and changing the access specifier from `PUBLIC RECORD`  to `PRIVATE FINAL RECORD`

I found no violation for HashSet so I removed  `// violation`

Rahulkhinchi03 replied:
After Adding Romani's code:

```
[ERROR] Failures: 
[ERROR]   IllegalTypeCheckTest.testArrayTypes:194->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeArrays.java differ.
missing (4): 20, 22, 25, 30
---
expected   : [20, 22, 24, 25, 29, 30]
but was    : [24, 29]
[ERROR]   IllegalTypeCheckTest.testGenerics:232->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestGenerics.java differ.
missing (5): 28, 29 [2 copies], 33, 68
---
expected   : [28, 29, 29, 32, 33, 33, 36, 36, 38, 40, 40, 42, 42, 46, 46, 55, 68]
but was    : [32, 33, 36, 36, 38, 40, 40, 42, 42, 46, 46, 55]
[ERROR]   IllegalTypeCheckTest.testIgnoreMethodNames:122->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestIgnoreMethodNames.java differ.
missing (2): 23, 26
---
expected   : [23, 26, 34, 43, 60, 62]
but was    : [34, 43, 60, 62]
[ERROR]   IllegalTypeCheckTest.testIllegalTypeEnhancedInstanceof:327->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources-noncompilable\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestEnhancedInstanceof.java differ.
missing (4): 28, 31, 35, 41
---
expected   : [28, 31, 35, 40, 41]
but was    : [40]
[ERROR]   IllegalTypeCheckTest.testIllegalTypeNewArrayStructure:356->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeNewArrayStructure.java differ.
missing (1): 26
---
expected   : [26]
but was    : []
[ERROR]   IllegalTypeCheckTest.testIllegalTypeRecordsAndCompactCtors:344->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources-noncompilable\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeRecordsAndCompactCtors.java differ.
missing (5): 27, 32, 35, 39, 48
---
expected   : [27, 31, 32, 35, 39, 40, 48]
but was    : [31, 40]
[ERROR]   IllegalTypeCheckTest.testLegalAbstractClassNames:153->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestLegalAbstractClassNames.java differ.
missing (1): 26
---
expected   : [26, 34, 35, 60, 62]
but was    : [34, 35, 60, 62]
[ERROR]   IllegalTypeCheckTest.testMemberModifiers:288->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestMemberModifiers.java differ.
missing (4): 22, 25, 39, 49
---
expected   : [22, 25, 32, 33, 39, 41, 49]
but was    : [32, 33, 41]
[ERROR]   IllegalTypeCheckTest.testPlainAndArrayTypes:207->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestPlainAndArraysTypes.java differ.
missing (3): 20, 24, 36
---
expected   : [20, 24, 26, 35, 36]
but was    : [26, 35]
[ERROR]   IllegalTypeCheckTest.testSameFileNameFalsePositive:164->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeSameFileNameFalsePositive.java differ.
missing (2): 28, 43
---
expected   : [28, 43]
but was    : []
[ERROR]   IllegalTypeCheckTest.testSameFileNameGeneral:180->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestSameFileNameGeneral.java differ.
missing (6): 25, 39, 40, 42, 43, 44
---
expected   : [25, 29, 31, 39, 40, 42, 43, 44]
but was    : [29, 31]
[ERROR]   IllegalTypeCheckTest.testStarImports:260->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestStarImports.java differ.
missing (1): 25
---
expected   : [25]
but was    : []
[ERROR]   IllegalTypeCheckTest.testStaticImports:272->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestStaticImports.java differ.
missing (1): 28
---
expected   : [28, 30]
but was    : [30]
[ERROR]   IllegalTypeCheckTest.testValidateAbstractClassNamesSetToTrue:52->AbstractModuleTestSupport.verifyWithInlineConfigParser:270->AbstractModuleTestSupport.verifyViolations:427 Violation lines for C:\Users\ASUS\Desktop\OpenSource\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\coding\illegaltype\InputIllegalTypeTestAbstractClassNamesTrue.java differ.
missing (1): 44
---
expected   : [27, 44, 46, 50]
but was    : [27, 46, 50]
[INFO] 
[ERROR] Tests run: 3652, Failures: 14, Errors: 0, Skipped: 1
```

Rahulkhinchi03 replied:
Should I resolve all these error or stick to only the current issue?

nmancus1 replied:
Update to check should look something like this: https://github.com/checkstyle/checkstyle/commit/959d778801a28d1e40d3f1bc268ec820660b1df0

Rahulkhinchi03 replied:
After doing this, there will be no violations, right?

Rahulkhinchi03 replied:
@nmancus1 After doing like #959d778 I am getting no violations for HashSet, is it okay or are we expecting any violation?

nmancus1 replied:
Amend all test cases to have comparable class definition for each given record definition to show identical violations (or lack thereof) as detailed in https://github.com/checkstyle/checkstyle/pull/10958/files#r758611207 . Also, please read about record definitions and record components to better understand how they relate to class definitions athttps://openjdk.java.net/jeps/359.

Rahulkhinchi03 replied:
@nmancus1  Please check my changes and Kindly reply for the same:
https://gist.github.com/Rahulkhinchi03/73db12098dc8d808b750ab3bc7d3c76e

nmancus1 replied:
You should amend test cases to include an equivalent class definition. This will help you to convince reviewers (and yourself) that:

```java

class TestClass {
    private final TreeSet x; // <- this declaration

    TestClass(TreeSet x) {
        this.x = x;
    }
}

```

is treated the same, with any configuration, that:

```java

record TestRecord(
        TreeSet x // <- this record component is
) {}

```

So for member modifiers like `LITERAL_PUBLIC, LITERAL_PROTECTED, LITERAL_STATIC`, the above examples should produce no violations, since `TestClass.x` is `private final` and `TestRecord.x` has implied modifiers `final` and `private`. The above examples would produce violations for the default config, and should for member modifiers like `FINAL` and `LITERAL_PRIVATE` or any combination thereof.

Rahulkhinchi03 replied:
Thank you so much @nmancus1  for the easy explanation.
@nmancus1 Please help me with the coverage issue: branch coverage is 0.97%

Rahulkhinchi03 replied:
Also, I am getting an error of `javac14` that I cannot use` private` in the private final record

nmancus1 replied:
To start with, you should have two or three test input files that are the same code, with different configs (this will be necessary for coverage), to show that violations (or lack thereof) are the same for test input that **includes** the following:

```java

import java.util.TreeSet;

class TestClass {
    private final TreeSet x; // <- any 'violation' or 'ok' here

    TestClass(TreeSet x) {
        this.x = x;
    }
}

record TestRecord(
        TreeSet x // // <- should be identical to `violation' or 'ok' here, for any config
) {}


```

Rahulkhinchi03 replied:
I am sorry but It is quite complex for be as a beginner

nmancus1 replied:
See https://github.com/nmancus1/checkstyle/commit/a50a96ffba8b8d8a0fa942bf4f1a7399b636a2f5

Rahulkhinchi03 replied:
Thanks a lot! @nmancus1 I understood now where my mistake was.

Rahulkhinchi03 replied:
branches covered ratio is 0.99, but expected minimum is 1.00

Rahulkhinchi03 replied:
I am still getting this issue @nmancus1 
`branches covered ratio is 0.99, but expected minimum is 1.00
`

Even after changing everything

nmancus1 replied:
Push your changes so that I can help; note that you do not need to modify the `isVerifiable` method in any way now, this method does not deal with record components at all anymore.

Rahulkhinchi03 replied:
> 

Using this code, I am getting violation error in file ` InputIllegalTypeRecordsAndCompactCtor`,

1. Should I use this code only and resolve these errors too?
2. Or Should I use: 
 ```
private void visitRecordComponentDef(DetailAST recordComponentDef) {
        if (memberModifiers.isEmpty()) {
            checkClassName(recordComponentDef);
        }
    }
```

Using point 2, I am getting no error.

Rahulkhinchi03 replied:
@nmancus1 I am again getting the branch and line overage issue. I only updated the Java Doc.

Rahulkhinchi03 replied:
@nmancus1  I am again getting the branch and line overage issue. I only updated the Java Doc and Added Romani sir's Code so I was getting an error and I updated to https://github.com/checkstyle/checkstyle/pull/10958#discussion_r762512351 

```
    /**
     * Checks a record component definition. Record components are private
     * final fields, so we do not consider memberModifiers field.
     *
     * @param recordComponentDef the record component definition to check.
     */
    private void visitRecordComponentDef(DetailAST recordComponentDef) {
            if (isContainVerifiableType(recordComponentDef)) {
                checkClassName(recordComponentDef);
            }
    }
```

To be precise:  `checkClassName(recordComponentDef);` This line is red and not in coverage

romani reviewed (changes requested):
Can we have input with violation on record?

Items:

Rahulkhinchi03 commented:
@romani sir, ping!

romani commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeRecordsWithMemberModifiersSpecified.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.UUID;
+
+public record InputIllegalTypeRecordsWithMemberModifiersSpecified
+        (HashSet productId, String identifier) {
```
This should be violation

romani replied:
Please add one more record absolutely same as in issue reported by user.

Rahulkhinchi03 replied:
Sure Sir.

romani reviewed (changes requested):
Please always add new test cases, the more district unique test cases the better.

Rahulkhinchi03 commented:
Github, rebase

romani commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeRecordsWithMemberModifiersSpecified.java:
```diff
@@ -0,0 +1,26 @@
+
+public record InputIllegalTypeRecordsWithMemberModifiersSpecified
+        (UUID productId, String identifier) {
+} // ok
```
Place comment on line where potential violation might be.
Previous line.

Rahulkhinchi03 replied:
Done Changes!

romani reviewed (changes requested):
Items:

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
@romani sir, please review!

romani commented:
Please do not resolve my review items, just reply "done" as you finish.

Rahulkhinchi03 commented:
\

> Please do not resolve my review items, just reply "done" as you finish.

Okay, I will do the same! Sorry for the inconvenience.

Rahulkhinchi03 commented:
@romani sir, all checks have been passed and I have resolved the changes.

nmancus1 commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeRecordsWithMemberModifiersSpecifiedTwo.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.*;
+
+public record InputIllegalTypeRecordsWithMemberModifiersSpecifiedTwo
+        (HashSet productId, String identifier) {  // violation
```
Record components are basically `private final` fields. Violations for the record components in this test case should be the same as the following for the given configs:

```
➜  javac Test1.java

➜  cat Test1.java
import java.util.HashSet;

public class Test1 {
    private final HashSet productId; // violation
    private final String identifier;

    public Test1(HashSet productId, String identifier) {
        this.productId = productId;
        this.identifier = identifier;
    }
}

➜  cat config.xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="IllegalType">
        </module>
    </module>
</module>                                                                                                                 

➜  java -jar checkstyle-9.1-all.jar -c config.xml Test1.java
Starting audit...
[ERROR] Test1.java:4:19: Usage of type 'HashSet' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 1 errors.

➜  cat config.xml 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="IllegalType">
                      <property name="memberModifiers" value="FINAL"/>
        </module>
    </module>
</module>                                                                                                                

➜  java -jar checkstyle-9.1-all.jar -c config.xml Test1.java
Starting audit...
[ERROR] Test1.java:4:19: Usage of type 'HashSet' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 1 errors.

➜  cat config.xml                                           
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="IllegalType">
                      <property name="memberModifiers" value="FINAL, LITERAL_PRIVATE"/>
        </module>
    </module>
</module>%                                                                                                                 ➜  company java -jar checkstyle-9.1-all.jar -c config.xml Test1.java
Starting audit...
[ERROR] Test1.java:4:19: Usage of type 'HashSet' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 1 errors.

```

The config from this test case should not have any violations:

```

➜  cat config.xml                                           
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="IllegalType">
                      <property name="memberModifiers" value="LITERAL_PUBLIC, LITERAL_PROTECTED, LITERAL_STATIC"/>
        </module>
    </module>
</module>                                                                                                               

➜  java -jar checkstyle-9.1-all.jar -c config.xml Test1.java
Starting audit...
Audit done.

```

Rahulkhinchi03 replied:
@nmancus1 I have changed the access specifier from 
```
public record InputIllegalTypeRecordsWithMemberModifiersSpecifiedTwo
        (HashSet productId, String identifier) {  // violation
}
```

-->

```
private final record InputIllegalTypeRecordsWithMemberModifiersSpecifiedTwo
        (HashSet productId,  // violation
         String identifier) {  
}
```

romani replied:
this is not exactly same as class fields.
Class fields can be HashSet but Ctor argument is Set:
```
public class Test1 {
    private final HashSet productId; // violation
    private final String identifier;

    public Test1(Set productId, String identifier) {
        this.productId = (HashSet)productId;
        this.identifier = identifier;
    }
}
```
so such Class does not expose HashSet requirements outside, and do "bad" stuff completely internally. So it is fine. 

"records" do expose illegal type outside and almost force all user to use illegal type.

We need to work/behave as on records as we behave on constructors.
if 
```
    public Test1(HashSet productId, String identifier) { // violation
        this.productId = (HashSet)productId;
        this.identifier = identifier;
    }
```
then
```
     record Test1(HashSet productId, String identifier); /// vilation
```

this should remove special treat consideration of FINAL and PRIVATE from code.

Rahulkhinchi03 replied:
@romani sir, I am getting no violation when I am using:
```
    public Test1(HashSet productId, String identifier) { // violation
        this.productId = (HashSet)productId;
        this.identifier = identifier;
    }
```

Rahulkhinchi03 replied:
Okay Done.

nmancus1 replied:
> We need to work/behave as on records as we behave on constructors.

This was my first thought as well, but this check does not check constructors parameters at all:

```java
    private void visitParameterDef(DetailAST parameterDef) {
        final DetailAST grandParentAST = parameterDef.getParent().getParent();

        if (grandParentAST.getType() == TokenTypes.METHOD_DEF && isCheckedMethod(grandParentAST)) {
            checkClassName(parameterDef);
        }
    }

```

Example:

```bash
        |--CTOR_DEF -> CTOR_DEF [4:4]
...
        |   |--PARAMETERS -> PARAMETERS [4:16]
        |   |   `--PARAMETER_DEF -> PARAMETER_DEF [4:16]
...
        |--METHOD_DEF -> METHOD_DEF [8:4]
...
        |   |--PARAMETERS -> PARAMETERS [8:16]
        |   |   `--PARAMETER_DEF -> PARAMETER_DEF [8:16]
...


```

This check relies on the given class's fields to put violations on illegal types, not constructor parameters. To me, this means that we should check record components as the fields they represent (`private final ...`), because this is how check currently operates on classes.

> so such Class does not expose HashSet requirements outside, and do "bad" stuff completely internally

This check does not have idea of "levels" of access modifiers; it only checks that if token has specified access modifiers (if `memberModifier` property is used) then we verify that it does not declare an illegal type. So, there is no concept of "outside" or "inside" to this check.

So, we have to make a decision on how to treat record components, and what implied modifiers that they carry. If we do not, they will not be checked at all if `memberModifiers` property is used at all. I can think of a few different solutions:

1. use example above (treat them as `private final`)
2. treat them as `public`
3. always check them, regardless of specified `memberModifiers` property as long as token is specified in config or default
4. let record components "inherit" access modifier of record declaration

Rahulkhinchi03 replied:
So Do I need to change anything? @nmancus1 @romani

Rahulkhinchi03 replied:
Do I need to revert changes again? @romani @nmancus1

Rahulkhinchi03 replied:
@romani sir, I updated the `SET in CTOR` and type cast `HashSet` in my recent changes.
https://github.com/checkstyle/checkstyle/pull/10958/files#diff-ca66a50cfe72a8a6786eac504d6efe1f7a3f6f79bb3428da3d07ae82330b4429R56-R58

romani replied:
> but this check does not check constructor parameters at all

it is separate new issue.
Check is very general, it tries to catch illegal classes in all places. It does validation of Method arguments. 

Please see proposal to update javadoc at https://github.com/checkstyle/checkstyle/pull/10958#discussion_r762465306 to make it clear that modifiers arenot considered for RECORD_COMPONENT_DEF

Rahulkhinchi03 replied:
@nmancus1 I think we can create a new issue regarding this?

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
Github, rebase

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java:
```diff
@@ -491,13 +493,13 @@ public void visitToken(DetailAST ast) {
-     * @param methodOrVariableDef METHOD_DEF or VARIABLE_DEF ast node.
+     * @param ast METHOD_DEF or VARIABLE_DEF ast node.
      * @return true if member is verifiable according to <b>memberModifiers</b> option.
      */
-    private boolean isVerifiable(DetailAST methodOrVariableDef) {
+    private boolean isVerifiable(DetailAST ast) {
         boolean result = true;
         if (!memberModifiers.isEmpty()) {
-            final DetailAST modifiersAst = methodOrVariableDef
+            final DetailAST modifiersAst = ast
```
Now these changes are not related to issue we are solving. So, you must either:

1. Revert these changes
2. Make these as a separate commit, something like `minor: refactor...`

I suggest (1), to help ease review in this PR; the less changes, the easier the review process, and increased confidence on reviewer's part of changes in contributor's PR.

Rahulkhinchi03 replied:
Okay, I am doing it.

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java:
```diff
@@ -569,10 +586,10 @@ private void visitParameterDef(DetailAST parameterDef) {
-     *
-     * @param variableDef variable to check.
-     */
+         * Checks type of given variable.
+         *
+         * @param variableDef variable to check.
+         */
```
Please revert this.

Rahulkhinchi03 replied:
Done Changes

romani replied:
Not done

Rahulkhinchi03 replied:
Sir, It is done and I am just pushing my code in a minute.

romani replied:
resolved

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
@romani sir and @nmancus1, please review my changes

romani commented:
@Rahulkhinchi03 , please consider https://github.com/checkstyle/checkstyle/pull/10958#discussion_r759963865

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
@romani sir, I cant understand the failing test cases: ` (Job MacOS JDK13 verify)`

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java:
```diff
@@ -525,6 +527,21 @@ private boolean isContainVerifiableType(DetailAST modifiers) {
+    private void visitRecordComponentDef(DetailAST recordComponentDef) {
+        if (memberModifiers.isEmpty()
+                || memberModifiers.contains(TokenTypes.LITERAL_PRIVATE)
+                || memberModifiers.contains(TokenTypes.FINAL)) {
```
I do not think this conditions are required.
See my comment above that it should be validated  like ctor

Rahulkhinchi03 replied:
@romani sir, should I push my previous changes CONDITIONS again instead of this?
Please checkout: https://github.com/Rahulkhinchi03/checkstyle/commit/7731e7cfa2c300bb36d4ba13f233075ca230b9e1

Rahulkhinchi03 replied:
Done!

Rahulkhinchi03 replied:
```
case TokenTypes.RECORD_COMPONENT_DEF:
                visitComponentDef(ast);
                break;
```

```
  /**
     * Checks type of ComponentDef.
     *
     * @param variableDef to check.
     */
    private void visitComponentDef(DetailAST variableDef) {
        if (isContainVerifiableType(variableDef)) {
            checkClassName(variableDef);
        }
    }
```
Using this code , I am getting violation error in file ` InputIllegalTypeRecordsAndCompactCtor`,



1. Should I use this code only and resolve these errors too?
2. Or Should I use: 
 ```
private void visitRecordComponentDef(DetailAST recordComponentDef) {
        if (memberModifiers.isEmpty()) {
            checkClassName(recordComponentDef);
        }
    }
```

Using point 2, I am getting no error.

romani replied:
see comment at https://github.com/checkstyle/checkstyle/pull/10958#discussion_r762465306

romani reviewed (changes requested):
Please fix CI

Items

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
@romani sir, please review my changes.

Rahulkhinchi03 commented:
@romani  sir, please review my changes.

Rahulkhinchi03 commented:
@romani sir, please review my changes.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java:
```diff
@@ -525,6 +527,19 @@ private boolean isContainVerifiableType(DetailAST modifiers) {
+     * @param recordComponentDef the record component definition to check.
+     */
+    private void visitRecordComponentDef(DetailAST recordComponentDef) {
+        if (memberModifiers.isEmpty()) {
```
this looks strange.
Lets remove this condition and update javadoc for this property `memberModifiers`. 
from `This property does not affect method calls nor method references. ` to `This property does not affect method calls nor method references nor record components. `

Rahulkhinchi03 replied:
@romani sir, I added your condition:
```
    /**
     * Checks a record component definition. Record components are private
     * final fields, so we do not consider memberModifiers field. This
     * property does not affect method calls nor method references
     * nor record components.
     *
     * @param variableDef the record component definition to check.
     */
    private void visitComponentDef(DetailAST variableDef) {
            if (isContainVerifiableType(variableDef)) {
                checkClassName(variableDef);
            }
    }
```

And all my test case for this Issue has been passed BUT I have got one error from another file then this issue:
File Name: InputIllegalTypeRecordsAndCompactCtors.java
Error:
expected   : [27, 31, 32, 35, 39, 40, 48]
but was    : [31, 32, 35, 40, 48]

Should I resolve these errors too? These are not related to this issue.

Rahulkhinchi03 replied:
Can we use this condition?
```
    private void visitRecordComponentDef(DetailAST recordComponentDef) {
        if (!memberModifiers.isEmpty()) {
            return;
        }
        checkClassName(recordComponentDef);
    }
```

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java:
```diff
@@ -525,6 +527,19 @@ private boolean isContainVerifiableType(DetailAST modifiers) {
+    /**
+     * Checks a record component definition. Record components are private
+     * final fields, so we check code in a similar manner as we do for
+     * private final class members.
```
```
* final fields, so we do not consider memberModifiers field.
```

Rahulkhinchi03 replied:
Done

Rahulkhinchi03 replied:
I will surely do it when I push changes again.

Rahulkhinchi03 replied:
Done

romani commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeRecordsWithMemberModifiersDefault.java:
```diff
@@ -0,0 +1,63 @@
+legalAbstractClassNames = (default)
+ignoredMethodNames = (default)getEnvironment, getInitialContext
+illegalAbstractClassNameFormat = (default)^(.*[.])?Abstract.*$
+memberModifiers = FINAL
```
it is not default, please make it default.

Rahulkhinchi03 replied:
Done

romani reviewed (changes requested):
please make a test without RCOND_COMPONENT_DEF in tokens to prove that no violation happens.

items:

Rahulkhinchi03 commented:
I updated the java Doc property for memberModifier and also removed the condition. Now, I am getting branches and lines coverage issues.
Specifically for these line: `checkClassName(recordComponentDef);`

Code:
```
    /**
     * Checks a record component definition. Record components are private
     * final fields, so we do not consider memberModifiers field. 
     *
     * @param variableDef the record component definition to check.
     */
    private void visitRecordComponentDef(DetailAST recordComponentDef) {
            if (isContainVerifiableType(recordComponentDef)) {
                checkClassName(recordComponentDef);
            }
    }
```

romani commented:
Please push your changes to GitHub to let me see them.

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
@romani sir, changes pushed.

Rahulkhinchi03 commented:
@romani  sir, changes pushed.

Rahulkhinchi03 commented:
@romani  sir, changes pushed.

Rahulkhinchi03 commented:
@romani sir, please review my changes.

romani commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeRecordsAndCompactCtors.java:
```diff
@@ -24,7 +24,7 @@
     record MyTestRecord
-            (LinkedHashMap<Integer, Integer> linkedHashMap) { // violation
+            (LinkedHashMap<Integer, Integer> linkedHashMap) { // OK
 
```
It should be violation

Rahulkhinchi03 replied:
I think the condition is not working with these violations.

romani replied:
We need to keep violation. Code should be updated.

Rahulkhinchi03 replied:
Can we use LinkedHashUp as a Parameter?

romani commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeRecordsWithMemberModifiersDefault.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    public record IdentifiersPair2(
+        HashSet x, // ok
```
It should be violation.
Why is it ok ?

Rahulkhinchi03 replied:
I first made it violation but I was getting an error that it is not violation with the given condition

Rahulkhinchi03 replied:
![44](https://user-images.githubusercontent.com/71710042/144758631-57bb0ba9-bc30-4f26-9b0a-9b58e8731ba3.png)

romani replied:
Please update expected list in Test.
It should be violation.

Rahulkhinchi03 replied:
45 is an expected list but it is still showing that it is not an error. I added the expected List in the Test but still it says that it shouldn't be the violation for this line.

Rahulkhinchi03 replied:
![aff](https://user-images.githubusercontent.com/71710042/144861831-e48b638a-bae4-44d5-a6f8-e9ca2e6b9832.png)

Rahulkhinchi03 replied:
I updated the ExpectedList in Test and also added  `HashSet x, // violation`
But I am still getting an error.

Rahulkhinchi03 replied:
@nmancus1 can you please help me with this?

nmancus1 replied:
Please see https://github.com/checkstyle/checkstyle/pull/10958#discussion_r762661072 

You need to make changes in the check to align with behavior specified by @romani at https://github.com/checkstyle/checkstyle/pull/10958#discussion_r762466240.

Rahulkhinchi03 replied:
@nmancus1 like this?
```
 private void visitRecordComponentDef(DetailAST recordComponentDef) {
        if (memberModifiers.contains(TokenTypes.RECORD_COMPONENT_DEF)) {
            checkClassName(recordComponentDef);
        }
    }
```

romani reviewed (changes requested):
Items:

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
@nmancus1 @romani sir,
I ran the coverage and found 100% but now it is showing branch coverage is 98% only.

Rahulkhinchi03 commented:
@romani sir, please review my changes!

Rahulkhinchi03 commented:
@romani  sir, please review my changes!

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
@romani sir,
I ran the coverage and found 100% but now it is showing branch coverage is 98% only. I updated the code. Please review my changes!

Rahulkhinchi03 commented:
Github, rebase

Rahulkhinchi03 commented:
@romani  sir,
I ran the coverage and found 100% but now it is showing branch coverage is 98% only. I updated the code. Please review my changes!

nmancus1 commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java:
```diff
@@ -525,6 +527,18 @@ private boolean isContainVerifiableType(DetailAST modifiers) {
+     * @param recordComponentDef the record component definition to check.
+     */
+    public void visitRecordComponentDef(DetailAST recordComponentDef) {
+        if (recordComponentDef.getType() == TokenTypes.RECORD_COMPONENT_DEF) {
```
Explain why we need this condition.

Rahulkhinchi03 replied:
We know that the RECORD_COMPONENT_DEF can catch illegal types from any place so we need a condition that can check to be it public, private, protected, final, We need a check which can check RECORD_COMPONENT-DEF type.

So if  whatever the type is of recordComponentDef, if it is equal to the token, we defined, then checkClassName(recordComponentDef);

nmancus1 replied:
Let me rephrase; how is this condition different than the precondition required to reach this method?

Rahulkhinchi03 replied:
In previous conditions, we were specifically checking for private and final member modifiers but Romani sir said that It not only check private and final but anything. So we have to consider them like constructors.

This condition is different from the pre-defined condition because before we were only checking the fields of the record components but now with this condition, we are checking every field and not specific to any particular field.

Rahulkhinchi03 replied:
So we had an option to how to treat these member modifiers and with your options that we can check public or private/final or we can check them for all the specified memberModifiers property

Rahulkhinchi03 replied:
Please guide me If It is wrong.

nmancus1 replied:
Please take some time, and read the documentation for `DetailAST#getType` and `TokenTypes`. Then, generate a coverage report https://github.com/checkstyle/checkstyle/wiki/How-to-run-certain-phases-and-validations to see what line is failing if you have not done this already. After that, try to understand why the code that is not covered is failing.

Rahulkhinchi03 replied:
Sure, thanks for the link.
@nmancus1 but every line is covered when I ran code coverage in my local environment. It's 100%

nmancus1 replied:
Then your local branch is not in sync with this PR branch.

Rahulkhinchi03 replied:
@nmancus1 I think this will work better?
 ```
 public void visitRecordComponentDef(DetailAST recordComponentDef) {
       checkClassName(recordComponentDef);
    }
```

nmancus1 replied:
Resolved

Rahulkhinchi03 commented:
Github, rebase

romani commented:
Github, rebase

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java:
```diff
@@ -525,6 +527,16 @@ private boolean isContainVerifiableType(DetailAST modifiers) {
+     * @param recordComponentDef the record component definition to check.
+     */
+    public void visitRecordComponentDef(DetailAST recordComponentDef) {
+        checkClassName(recordComponentDef);
```
Simply use this method in "case"

Rahulkhinchi03 replied:
So I should delete this method here and update it in case?

Rahulkhinchi03 replied:
But in case we are using (DetailAST ast) so how can I use (DetailASR recordComponentDef) in CASE?

Rahulkhinchi03 replied:
@romani sir, Can we use this condition?
```
public void visitRecordComponentDef(DetailAST recordComponentDef) {
        if (recordComponentDef.getType() == TokenTypes.RECORD_COMPONENT_DEF) {
            checkClassName(recordComponentDef);
        }
```

romani replied:
```
case TokenTypes.RECORD_COMPONENT_DEF:
                checkClassName(ast);
```

Rahulkhinchi03 replied:
Thanks Sir!
Should I delete this method now?

Rahulkhinchi03 replied:
Done!

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheckTest.java:
```diff
@@ -330,7 +329,6 @@ public void testIllegalTypeRecordsAndCompactCtors() throws Exception {
 
     @Test
     public void testIllegalTypeNewArrayStructure() throws Exception {
-
```
Restore changes unrelated to your issue

Rahulkhinchi03 replied:
Done!

romani commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeRecordsAndCompactCtors.java:
```diff
@@ -36,7 +36,7 @@ record MyTestRecord2(String string) implements Cloneable { // violation
 
-    record MyTestRecord3(String str, TreeSet treeSet) { // violation
+    record MyTestRecord3(String str, Set treeSet) { // OK
         void foo(HashMap<Integer, Integer> hashMap) { // violation
```
Restore this line, it should not be charged

Rahulkhinchi03 replied:
Done!

romani reviewed (changes requested):
Items:

Rahulkhinchi03 commented:
Github, rebase

romani reviewed (approved):
Ok by me.
I extended list of reviewers, as PR was not clear .

Rahulkhinchi03 commented:
Github, rebase

nmancus1 commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeRecordsWithMemberModifiersDefault.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    public record IdentifiersPair2(
+        HashSet x, // violation
```
@romani what is the policy for violation messages in input files? Do we require violation messages on all, or only when conditions are more complex or check has multiple messages?

romani replied:
When message could be different from Check.
If Check has single message - no need to print it. But no harm to have it if scecific  case is confusing

Rahulkhinchi03 replied:
@romani sir, I had one doubt.
You said first that we can only use `Set` in CTOR and not `HashSet` but then you said me to revert changes. wasn't that `// ok`

romani replied:
We can use any type at any place.
Check should detect illegally config type at required place and place violation on it.

In test we need all possible combinations.
So do not touch existing tests and just deal with existing.

Rahulkhinchi03 commented:
@pbludov please review my changes!

Rahulkhinchi03 commented:
@strkkk kindly review my changes!
